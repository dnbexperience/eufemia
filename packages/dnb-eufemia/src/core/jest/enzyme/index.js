/**
 * Default Component Test Setup
 *
 */

// enzyme setup
const enzyme = require('enzyme')

// TODO: This may not be needed in future
const Adapter = require('enzyme-adapter-react-16')
enzyme.configure({ adapter: new Adapter() })

module.exports = enzyme
