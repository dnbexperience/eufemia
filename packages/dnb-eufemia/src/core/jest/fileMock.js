export default 'static file content'
