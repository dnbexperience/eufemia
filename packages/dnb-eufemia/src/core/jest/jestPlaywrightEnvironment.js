/**
 * Jest Setup for Screenshot testing
 *
 */

const PlaywrightEnvironment =
  require('jest-playwright-preset/lib/PlaywrightEnvironment').default
const chalk = require('chalk')
const { config } = require('./jestSetupScreenshots')

class JestEnvironment extends PlaywrightEnvironment {
  constructor(config, context) {
    super(config, context)
  }

  async setup() {
    await super.setup()

    // jest-playwright exposes these on the global
    const browser = this.global.browser
    const browserName = this.global.browserName || 'firefox'

    if (browser && typeof browser.version === 'function') {
      const version = browser.version() // e.g. "129.0"
      // One neat line before your test output:
      console.log(`\nâ–¶ Using ${browserName} ${version}\n`)
    }
  }

  async teardown() {
    await super.teardown()
  }

  setRetryAttempt(retryAttempt) {
    this.global.retryAttempt = retryAttempt
  }

  getCurrentTestName(state) {
    const { currentlyRunningTest } = state
    return getParents(currentlyRunningTest).reverse().join(' ')
  }

  reportedPageUrl = []

  async handleTestEvent(event, state) {
    if (
      this.global.pageUrl &&
      !this.reportedPageUrl.includes(this.global.pageUrl)
    ) {
      this.reportedPageUrl.push(this.global.pageUrl)

      const cliColors = {
        reset: '\x1b[0m',
        bold: '\x1b[1m',
        dim: '\x1b[2m',
        yellow: '\x1b[33m',
        red: '\x1b[31m',
        green: '\x1b[32m',
        hidden: '\x1b[8m',
      }

      const themeName = this.global.themeName
        ? `${cliColors.bold}${this.global.themeName}${cliColors.reset} `
        : ''
      console.log(
        `${cliColors.yellow}URL:`,
        `${cliColors.reset}${themeName}${this.global.pageUrl}`
      )
    }

    if (config.retryTimes > 0) {
      // Reset retry attempt at the start of each test
      if (event.name === 'test_start') {
        this.setRetryAttempt(0)
      }

      if (event.name === 'test_fn_failure') {
        const currentTestName = this.getCurrentTestName(state)
        const retryAttempt = state.currentlyRunningTest.invocations
        this.setRetryAttempt(retryAttempt)

        // Set a marker on the page to indicate this is a retry
        if (this.global.page) {
          this.global.page
            .evaluate(() => {
              window.__VISUAL_TEST_RETRY__ = true
            })
            .catch(() => {
              // Ignore errors if page is not available
            })
        }

        console.log(
          chalk.yellow(
            `Retry attempt #${retryAttempt}: ${currentTestName}`
          )
        )
      }

      if (event.name === 'test_fn_success') {
        // Reset retry attempt on success
        this.setRetryAttempt(0)
      }
    }
  }
}

module.exports = JestEnvironment

function getParents(item) {
  const names = []

  if (item?.name) {
    names.push(item.name)
  }

  for (const key in item) {
    if (item[key]?.name) {
      names.push(item[key].name)
    }
    if (item[key]?.parent) {
      names.push(...getParents(item[key].parent))
    }
  }

  return names.filter(
    (name) => name !== 'ROOT_DESCRIBE_BLOCK' && name !== 'Error'
  )
}
