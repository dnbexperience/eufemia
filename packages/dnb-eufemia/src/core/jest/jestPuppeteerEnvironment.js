/**
 * Jest Setup for Screenshot testing
 *
 */

const PlaywrightEnvironment =
  require('jest-playwright-preset/lib/PlaywrightEnvironment').default
const chalk = require('chalk')
const { config } = require('./jestSetupScreenshots')
const { slugify } = require('../../shared/component-helper')

class JestEnvironment extends PlaywrightEnvironment {
  constructor(config, context) {
    super(config, context)
  }

  async setup() {
    await super.setup()
  }

  async teardown() {
    await super.teardown()
  }

  getCurrentTestName(state) {
    const { currentlyRunningTest } = state
    return getParents(currentlyRunningTest).reverse().join(' ')
  }

  reportedPageUrl = []

  async handleTestEvent(event, state) {
    if (
      this.global.pageUrl &&
      !this.reportedPageUrl.includes(this.global.pageUrl)
    ) {
      this.reportedPageUrl.push(this.global.pageUrl)

      const cliColors = {
        reset: '\x1b[0m',
        bold: '\x1b[1m',
        dim: '\x1b[2m',
        yellow: '\x1b[33m',
        red: '\x1b[31m',
        green: '\x1b[32m',
        hidden: '\x1b[8m',
      }

      const themeName = this.global.themeName
        ? `${cliColors.bold}${this.global.themeName}${cliColors.reset} `
        : ''
      console.log(
        `${cliColors.yellow}URL:`,
        `${cliColors.reset}${themeName}${this.global.pageUrl}`
      )
    }

    if (config.retryTimes > 0) {
      this.global.retryAttempt = 0

      if (event.name === 'test_fn_failure') {
        const currentTestName = this.getCurrentTestName(state)
        const slug = slugify(currentTestName)

        if (typeof global.__EVENT_FAILURE_CACHE__ === 'undefined') {
          global.__EVENT_FAILURE_CACHE__ = {}
        }
        global.__EVENT_FAILURE_CACHE__[slug] = {
          currentTestName,
          failed: true,
        }

        const retryAttempt = state.currentlyRunningTest.invocations
        this.global.retryAttempt = retryAttempt

        console.log(
          chalk.yellow(
            `Retry attempt #${retryAttempt}: ${currentTestName}`
          )
        )
      }

      if (event.name === 'test_fn_success') {
        const currentTestName = this.getCurrentTestName(state)

        const slug = slugify(currentTestName)
        if (global.__EVENT_FAILURE_CACHE__[slug]) {
          global.__EVENT_FAILURE_CACHE__[slug].failed = false
        }
      }
    }
  }
}

module.exports = JestEnvironment

function getParents(item) {
  const names = []

  if (item?.name) {
    names.push(item.name)
  }

  for (const key in item) {
    if (item[key]?.name) {
      names.push(item[key].name)
    }
    if (item[key]?.parent) {
      names.push(...getParents(item[key].parent))
    }
  }

  return names.filter(
    (name) => name !== 'ROOT_DESCRIBE_BLOCK' && name !== 'Error'
  )
}
