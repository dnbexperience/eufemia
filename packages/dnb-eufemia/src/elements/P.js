/**
 * HTML Element
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { spacingPropTypes } from '../components/space/SpacingHelper'
import E from './Element'

const P = ({
  modifier,
  element,
  className,
  small,
  medium,
  bold,
  size,

  ...props
}) => {
  if (typeof modifier === 'string' && / /.test(modifier)) {
    modifier = modifier.split(/ /g)
  } else if (!Array.isArray(modifier)) {
    modifier = [modifier]
  }

  if (medium === true) {
    modifier.push('medium')
  } else if (bold === true) {
    modifier.push('bold')
  }

  modifier = modifier.filter(Boolean).reduce((acc, cur) => {
    if (['x-small', 'small'].includes(cur)) {
      return `${acc} dnb-p__size--${cur}`
    }
    return `${acc} dnb-p--${cur}`
  }, '')

  if (size) {
    className = classnames(className, `dnb-p__size--${size}`)
  } else if (small === true) {
    className = classnames(className, 'dnb-p__size--small')
  }

  return (
    <E
      as={element}
      {...props}
      className={classnames('dnb-p', modifier, className)}
    />
  )
}
P.tagName = 'dnb-p'
P.propTypes = {
  ...spacingPropTypes,

  element: PropTypes.node,
  className: PropTypes.string,
  small: PropTypes.bool,
  medium: PropTypes.bool,
  bold: PropTypes.bool,
  size: PropTypes.oneOf([
    'x-small',
    'small',
    'basis',
    'medium',
    'large',
    'x-large',
    'xx-large',
  ]),
  modifier: PropTypes.string,
  children: PropTypes.node,
}
P.defaultProps = {
  element: 'p',
  className: null,
  small: null,
  medium: null,
  bold: null,
  size: null,
  modifier: null,
  children: null,
}
export default P
