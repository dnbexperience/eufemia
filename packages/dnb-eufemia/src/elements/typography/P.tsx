/**
 * HTML Element
 *
 */

import React from 'react'
import classnames from 'classnames'
import { SpacingProps } from '../../components/space/types'
import type { DynamicElement } from '../../shared/types'
import E from '../Element'

export type PSize =
  | 'x-small'
  | 'small'
  | 'basis'
  | 'medium'
  | 'large'
  | 'x-large'
  | 'xx-large'

export type PProps = SpacingProps &
  React.HTMLAttributes<HTMLParagraphElement> & {
    /**
     * Defines the Element Type, like "p"
     * Default: p
     */
    element?: DynamicElement & 'p'
    /**
     * Tells the component to use the medium font-weight styling dnb-p--medium defined in paragraphStyle - typography-mixins.scss. Find more details here https://eufemia.dnb.no/uilib/typography/font-weight/
     */
    medium?: boolean
    /**
     * Tells the component to use the bold font-weight styling dnb-p--bold defined in paragraphStyle - typography-mixins.scss. Find more details here https://eufemia.dnb.no/uilib/typography/font-weight/
     */
    bold?: boolean
    /**
     * Sets the font size based on size classes defined in paragraphStyle - typography-mixins.scss. For more detailed information go here: https://eufemia.dnb.no/uilib/typography/font-size/
     */
    size?: PSize
    /**
     * A string containing a combination of modifiers, used to set both font-size and weight in one property. e.g. "x-small bold" would make the paragraph extra small and bold.
     * Works as a flexible alternative to setting the medium, small, bold and size props.
     * List of modifiers can be found at https://eufemia.dnb.no/uilib/typography/font-size/ and https://eufemia.dnb.no/uilib/typography/font-weight/
     */
    modifier?: string
  }

const P = ({
  modifier,
  element = 'p',
  className,
  medium,
  bold,
  size,
  ...props
}: PProps) => {
  const allModifiers = [medium && 'medium', bold && 'bold']

  if (modifier) {
    modifier
      .split(/\s/g)
      .forEach((modifier) => allModifiers.push(modifier))
  }

  const modifierString = allModifiers
    .filter(Boolean)
    .reduce((acc, cur) => {
      if (['x-small', 'small'].includes(cur)) {
        return `${acc} dnb-p__size--${cur}`
      }

      return `${acc} dnb-p--${cur}`
    }, '')

  return (
    <E
      as={element}
      {...props}
      className={classnames(
        'dnb-p',
        modifierString,
        className,
        size && `dnb-p__size--${size}`
      )}
    />
  )
}

export default P
