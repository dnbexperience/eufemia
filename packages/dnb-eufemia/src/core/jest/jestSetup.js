/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
