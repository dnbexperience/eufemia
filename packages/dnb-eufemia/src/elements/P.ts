/**
 * This file is created manually
 * You can change the content
 *
 */

import P from './typography/P'
export * from './typography/P'
export default P
