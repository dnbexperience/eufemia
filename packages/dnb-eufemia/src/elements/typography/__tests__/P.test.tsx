/**
 * Element Test
 *
 */

import React from 'react'
import { axeComponent } from '../../../core/jest/jestSetup'
import P, { PProps } from '../P'
import { render } from '@testing-library/react'

const props: PProps = {
  size: 'x-small',
  element: 'p',
}

describe('P element', () => {
  it('has correct size when size is defined', () => {
    render(<P size="large" />)
    const element = document.querySelector('.dnb-p__size--large')

    expect(Array.from(element.classList)).toEqual([
      'dnb-p',
      'dnb-p__size--large',
    ])
  })
  it('has correct style when size and a modifier is defined', () => {
    render(<P size="medium" modifier="medium" />)
    const element = document.querySelector('.dnb-p__size--medium')

    expect(Array.from(element.classList)).toEqual([
      'dnb-p',
      'dnb-p--medium',
      'dnb-p__size--medium',
    ])
  })
  it('has correct style when several modifiers are defined', () => {
    render(<P modifier="medium small" />)
    const element = document.querySelector('.dnb-p__size--small')

    expect(Array.from(element.classList)).toEqual([
      'dnb-p',
      'dnb-p--medium',
      'dnb-p__size--small',
    ])
  })
  it('has correct style when medium is set to true', () => {
    render(<P medium />)
    const element = document.querySelector('.dnb-p--medium')
    expect(Array.from(element.classList)).toEqual([
      'dnb-p',
      'dnb-p--medium',
    ])
  })
  it('has correct style when bold is set to true', () => {
    render(<P bold />)
    const element = document.querySelector('.dnb-p--bold')

    expect(Array.from(element.classList)).toEqual(['dnb-p', 'dnb-p--bold'])
  })
  it('should validate with ARIA rules as a p element', async () => {
    const Comp = render(<P {...props} />)
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})
