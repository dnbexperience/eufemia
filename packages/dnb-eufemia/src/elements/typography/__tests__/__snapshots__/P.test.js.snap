// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`P element have to match default P element snapshot 1`] = `
<P
  bold={true}
  element="p"
  medium={true}
  modifier="Error, unknown type"
  size="x-small"
  small={true}
>
  <ForwardRef
    as="p"
    className="dnb-p  dnb-p--Error, dnb-p--unknown dnb-p--type dnb-p--medium dnb-p__size--x-small dnb-p__size--small"
  >
    <ElementInstance
      as="p"
      className="dnb-p  dnb-p--Error, dnb-p--unknown dnb-p--type dnb-p--medium dnb-p__size--x-small dnb-p__size--small"
      innerRef={null}
    >
      <p
        className="dnb-p  dnb-p--Error, dnb-p--unknown dnb-p--type dnb-p--medium dnb-p__size--x-small dnb-p__size--small"
      />
    </ElementInstance>
  </ForwardRef>
</P>
`;
